File: NSNumber+Times.m
#import "NSNumber+Times.h"

@interface NSNumber (timesimplementation)
-realTimes:target perform:(SEL)action withObject:object;
-nonZero;
@end

@implementation NSNumber (times)

-times:target perform:(SEL)action
{
    return [self times:target perform:action withObject:nil];
}

-times:target perform:(SEL)action withObject:object
{
    return [[self nonZero] realTimes:target
                             perform:action
                          withObject:object];
}

@end

@implementation NSNumber (timesimplementation)

-realTimes:target perform:(SEL)action withObject:object
{
    [target performSelector:action withObject:object];
    return [[NSNumber numberWithInteger:[self integerValue] - 1]
            times:target perform:action withObject:object];
}

-nonZero
{
    return ([self integerValue] != 0)?self:nil;
}

@end

File: Life/Swift/NSNumber+Times.swift
//
//  NSNumber+Times.swift
//  Life
//
//  Created by Nityananda Zbil on 29.07.20.
//  Copyright © 2020 Nityananda Zbil. All rights reserved.
//

import Foundation

@objc
extension NSNumber {
    func times(_ target: Any!, perform action: Selector!) -> Any! {
        return times(target, perform: action, withObject: nil)
    }
    
    func times(_ target: Any!, perform action: Selector!, withObject object: Any!) -> Any! {
        return nonZero()?.realTimes(target, perform: action, with: object)
    }
}

private extension NSNumber {
    func realTimes(_ target: Any!, perform action: Selector!, with object: Any!) -> Any! {
        (target as? NSObject)?.perform(action, with: object)
        return NSNumber(integerLiteral: intValue - 1)
            .times(target, perform: action, withObject: object)
    }
    
    func nonZero() -> NSNumber? {
        intValue != 0
            ? self
            : nil
    }
}
